### PR TITLE
Get `$nu.config-path` and `$nu.env-path` from `EngineState`

### LINKED ISSUE #6344: Custom `config.nu` location doesn't change `$nu.config-path`
### Describe the bug

in windows, starting nu with `--config <custom-config-location>`  doesn't change  `$nu.config-path` to custom config path.

### How to reproduce

1. run nu with custom config location `nu.exe --config <custom-config-location> --env-config <custom-env-location>`
2. type `$nu.config-path` in nushell

### Expected behavior

the result of `$nu.config-path` should be the path of custom config.nu

### Screenshots

image

### Configuration

| key                | value                                           |
| ------------------ | ----------------------------------------------- |
| version            | 0.67.0                                          |
| branch             |                                                 |
| commit_hash        | 33674d3a98dd7510a4e4784c079dfdaecfd2e0d6        |
| build_os           | windows-x86_64                                  |
| build_target       | x86_64-pc-windows-msvc                          |
| rust_version       | rustc 1.63.0 (4b91a6ea7 2022-08-08)             |
| rust_channel       | stable-x86_64-pc-windows-msvc                   |
| cargo_version      | cargo 1.63.0 (fd9c4297c 2022-07-01)             |
| pkg_version        | 0.67.0                                          |
| build_time         | 2022-08-16 20:28:43 +00:00                      |
| build_rust_channel | release                                         |
| features           | database, dataframe, default, trash, which, zip |
| installed_plugins  |                                                 |

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by nibon7**
# Description

Fixes #6344

Screenshot from 2022-08-22 15-36-01

# Tests

Make sure you've done the following:

- [ ] Add tests that cover your changes, either in the command examples, the crate/tests folder, or in the /tests folder.
- [ ] Try to think about corner cases and various ways how your changes could break. Cover them with tests.
- [ ] If adding tests is not possible, please document in the PR body a minimal example with steps on how to reproduce so one can verify your change works.

Make sure you've run and fixed any issues with these commands:

- [x] `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- [x] `cargo clippy --workspace --features=extra -- -D warnings -D clippy::unwrap_used -A clippy::needless_collect` to check that you're using the standard code style
- [x] `cargo test --workspace --features=extra` to check that all the tests pass

### CODE REVIEW

**Comment by fdncred** [comment]
I don't understand why these need to be environment variables in order to make this change. Please explain.

**Comment by nibon7** [comment]
> I don't understand why these need to be environment variables in order to make this change. Please explain.

I didn't find a better way to fix the issue, and seems environment variables works :). `eval_variable` doesn't care about `--config` and `--env-config`, so I use environment variables to pass the `config-path` and `env-path` to `eval_variables`.

**Comment by nibon7** [comment]
> And lastly, what about the login.nu file?

`nu` doesn't provide an option to load alternate `login.nu`, I think it's ok.

**Comment by fdncred** [comment]
Could someone just set the env vars to change the location of the config files instead of passing them in with --config and --env-config?

**Comment by nibon7** [comment]
> Could someone just set the env vars to change the location of the config files instead of passing them in with --config and --env-config?

They can be changed, just like `NU_LIB_DIRS` and `NU_PLUGIN_DIRS`.

**Comment by rgwood** [comment]
Hey @nibon7, I tried to review this PR but it's a little hard to follow how the config+env paths are set now.

We really appreciate the effort, but could you please add more explanations to your PRs? It would be helpful to have short summaries of things like:

- what problem this solves
- how things worked before the PR
- how things work after the PR

**Comment by nibon7** [comment]
> Hey @nibon7, I tried to review this PR but it's a little hard to follow how the config+env paths are set now.
> 
> We really appreciate the effort, but could you please add more explanations to your PRs? It would be helpful to have short summaries of things like:
> 
>     * what problem this solves
> 
>     * how things worked before the PR
> 
>     * how things work after the PR

`$nu.config-path` or `$nu.env-path` always prints `$HOME/.config/nushell/config.nu` or  `$HOME/.config/nushell/env.nu` before this PR, this is incorrect when user provides `--config /some/path/to/config.nu` or `--env-config /some/path/to/env.nu`. This PR saves `config-path` or `env-path` to environment variables before we sourcing `config.nu` or `env.nu`, when we print `$nu.config-path` or `$nu.env-path`, we get them from those environment variables, not fixed path like `$HOME/.config/nushell/xxx.nu`

**Comment by kubouch** [comment]
The main problem I see is that you can change the environment variables which would crate a situation when they're pointing at different files. `NU_LIB_DIRS` and `NU_PLUGIN_DIRS` are supposed to be changed and their change is reflected in the code so it's a different case.

I'd rather add the paths to the existing `$nu.xxx-path` variables. The `$nu` is created dynamically on-request so we need to store the paths somewhere, probably the Config struct stored in the EngineState. You could create a `set_config_path()` method to the Config struct that would set these paths. During eval, when constructing the `$nu` variable, the paths would be grabbed from there.

**Comment by nibon7** [comment]
> The `$nu` is created dynamically on-request so we need to store the paths somewhere, probably the Config struct stored in the EngineState

Would it be better to store the paths directly to `EngineState`? Seems when `config.nu` is loaded, the `Config` items will be overridden.

**Comment by kubouch** [comment]
Thanks, yeah, I think it should be ok. I would make one small change: If `get_config_file()` lookup fails, it should put the default `config.nu` or `env.nu` location. So if you don't specify them with arguments, `$nu.config-path` and `$nu.env-path` should point at the default location.

It would also be great to add tests for this or at least write up a minimal working example that demonstrates how it works (the screenshot is outdated now).

**Comment by nibon7** [comment]
> If `get_config_file()` lookup fails, it should put the default `config.nu` or `env.nu` location.

```diff
@@ -1321,21 +1337,25 @@ pub fn eval_variable(
                     span,
                 });

-                config_path.push("config.nu");
+                if engine_state.get_config_path("config-path").is_none() {
+                    config_path.push("config.nu");

-                output_cols.push("config-path".into());
-                output_vals.push(Value::String {
-                    val: config_path.to_string_lossy().to_string(),
-                    span,
-                });
+                    output_cols.push("config-path".into());
+                    output_vals.push(Value::String {
+                        val: config_path.to_string_lossy().to_string(),
+                        span,
+                    });
+                }

-                env_config_path.push("env.nu");
+                if engine_state.get_config_path("env-path").is_none() {
+                    env_config_path.push("env.nu");

-                output_cols.push("env-path".into());
-                output_vals.push(Value::String {
-                    val: env_config_path.to_string_lossy().to_string(),
-                    span,
-                });
+                    output_cols.push("env-path".into());
+                    output_vals.push(Value::String {
+                        val: env_config_path.to_string_lossy().to_string(),
+                        span,
+                    });
+                }

                 loginshell_path.push("login.nu");
```

**Comment by nibon7** [comment]
> So if you don't specify them with arguments, `$nu.config-path` and `$nu.env-path` should point at the default location.

If user doesn't specify `--config` or `--env-config`, `config_file` or `env_file` will be `None`, and it will looks up the default `config.nu` or `enu.nu`.

```diff
@@ -690,3 +700,36 @@ fn set_is_perf_value(value: bool) {
         *new_value.borrow_mut() = value;
     });
 }
+
+fn set_config_path(
+    engine_state: &mut EngineState,
+    cwd: &Path,
+    config_file: &Option<Spanned<String>>,
+    env_file: &Option<Spanned<String>>,
+) {
+    let config_path = match config_file {
+        Some(s) => canonicalize_with(&s.item, cwd).ok(),
+        None => nu_path::config_dir().map(|mut p| {
+            p.push(config_files::NUSHELL_FOLDER);
+            p.push("config.nu");
+            p
+        }),
+    };
+
+    if let Some(path) = config_path {
+        engine_state.set_config_path("config-path", path);
+    }
+
+    let env_path = match env_file {
+        Some(s) => canonicalize_with(&s.item, cwd).ok(),
+        None => nu_path::config_dir().map(|mut p| {
+            p.push(config_files::NUSHELL_FOLDER);
+            p.push("env.nu");
+            p
+        }),
+    };
+
+    if let Some(path) = env_path {
+        engine_state.set_config_path("env-path", path);
+    }
+}
```

**Comment by nibon7** [comment]
> It would also be great to add tests for this or at least write up a minimal working example that demonstrates how it works (the screenshot is outdated now).

Seems it's hard to write test cases using the existing test framework, since we need to pass `--config` or `--env-config` to `nu!` and `nu!` doesn't take extra args.  I've updated the screenshot.

**Comment by fdncred** [comment]
thanks for changing this @nibon7. it looks much better. we'll wait to see what kubouch says about tests before we land this.

**Comment by kubouch** [comment]
Ah, yeah, I didn't see the `is_none()` checks below. Looks good to me. Thanks for fixing this!

And yes, the tests with CLI args are a bit complicated. However, I think calling `nu --config=...` inside the `nu!` macro should work. Check how `nu_repl_code()` (or testbins in general) is used in tests. I'll merge it because the change is relatively simple but if you feel like adding the tests or tweaking our test suite to be more ergonomic with CLI arguments, please go ahead.

**Comment by nibon7** [comment]
> Check how `nu_repl_code()` (or testbins in general) is used in tests

Seems `nu!(cwd: dirs.next(),  nu_repl_code(inp_lines)` start a new `nu` process.

https://github.com/nushell/nushell/blob/7ef4e5f9401469888a37f6fdd592f65a67cd7610/crates/nu-test-support/src/lib.rs#L39-L56
